Source/DafnyLanguageServer/Handlers/DafnyTextDocumentHandler.cs

```diff
@@ -65,102 +61,54 @@ public class DafnyTextDocumentHandler : TextDocumentSyncHandlerBase {
       logger.LogTrace("received save notification {DocumentUri}", notification.TextDocument.Uri);
-      ForwardDiagnostics(notification.TextDocument.Uri, documents.SaveDocument(notification.TextDocument));
-      return Unit.Task;
-    }
-
-    private void ForwardDiagnostics(DocumentUri uri, IObservable<DafnyDocument> requestUpdates) {
-      var observer = observers.GetOrCreate(uri, () => new RequestUpdatesOnUriObserver(logger, telemetryPublisher, diagnosticPublisher));
-      observer.OnNext(requestUpdates);
-    }
-
-    private class RequestUpdatesOnUriObserver : IObserver<IObservable<DafnyDocument>>, IDisposable {
-      private readonly MergeOrdered<DafnyDocument> mergeOrdered;
-      private readonly IDisposable subscription;
-
-      public RequestUpdatesOnUriObserver(ILogger logger, ITelemetryPublisher telemetryPublisher,
-        IDiagnosticPublisher diagnosticPublisher) {
-
-        mergeOrdered = new MergeOrdered<DafnyDocument>();
-        subscription = mergeOrdered.Subscribe(new DiagnosticsObserver(logger, telemetryPublisher, diagnosticPublisher));
-      }
-
-      public void Dispose() {
-        subscription.Dispose();
-      }
-
-      public void OnCompleted() {
-        mergeOrdered.OnCompleted();
-      }
-
-      public void OnError(Exception error) {
-        mergeOrdered.OnError(error);
-      }
-
-      public void OnNext(IObservable<DafnyDocument> value) {
-        mergeOrdered.OnNext(value);
-      }
-    }
-
-    private class DiagnosticsObserver : IObserver<DafnyDocument> {
-      private readonly ILogger logger;
-      private readonly ITelemetryPublisher telemetryPublisher;
-      private readonly IDiagnosticPublisher diagnosticPublisher;
-
-      public DiagnosticsObserver(ILogger logger, ITelemetryPublisher telemetryPublisher, IDiagnosticPublisher diagnosticPublisher) {
-        this.logger = logger;
-        this.telemetryPublisher = telemetryPublisher;
-        this.diagnosticPublisher = diagnosticPublisher;
-      }
-
-      public void OnCompleted() {
-        telemetryPublisher.PublishUpdateComplete();
-      }
-
-      public void OnError(Exception e) {
-        if (e is TaskCanceledException) {
-          OnCompleted();
-        } else {
-          logger.LogError(e, "error while handling document event");
-          telemetryPublisher.PublishUnhandledException(e);
-        }
+      try {
+        documents.SaveDocument(notification.TextDocument);
+      } catch (Exception e) {
+        telemetryPublisher.PublishUnhandledException(e);
       }
 
-      public void OnNext(DafnyDocument document) {
-        diagnosticPublisher.PublishDiagnostics(document);
-      }
+      return Unit.Task;
     }
 
     private async Task CloseDocumentAndHideDiagnosticsAsync(TextDocumentIdentifier documentId) {
       try {
         await documents.CloseDocumentAsync(documentId);
+        notificationPublisher.HideDiagnostics(documentId);
```

Comment:
It's already hiding diagnostics in VSCode when I close a file, but I guess here it's more implementing this behavior on the LSP protocol side of view?

Reply:
I'm not sure, but I moved this code from a few lines below, so this change is just about maintaining the existing behavior.